source-oracle/source_oracle/api.py

```diff
@@ -0,0 +1,332 @@
+    full_state: ConnectorState,
+) -> AsyncGenerator[Document | LogCursor, None]:
+    # fetch_changes only runs on odd sync counter numbers
+    # this allows a ping-ponging between fetch_page and fetch_changes
+    if full_state.sync_counter % 2 != 1 and full_state.backfill is not None:
+        return
+
+    scn = log_cursor
+    query = template_env.get_template("inc").render(table=table, cursor=scn)
+
+    log.debug("fetch_changes", query, log_cursor)
+
+    # We may receive many documents from the same transaction, as such we need to only emit the checkpoint
+    # for a transaction at the end, when all documents of that transaction have been emitted
+    # so we have to keep track of the first two transaction SCNs we have to check whether we have all the documents
+    # of the current SCN before moving on to the next one.
+    first_transaction = None
+    current_transaction = None
+    async with pool.acquire() as conn:
+        with conn.cursor() as c:
+            await c.execute(query)
+            async for values in c:
+                cols = [col[0] for col in c.description]
+                row = dict(zip(cols, values))
+                row = {k: v for (k, v) in row.items() if v is not None}
+
+                scn = row[scn_column_name]
+                op = row[op_column_name]
+                row_id = row[rowid_column_name]
+
+                doc = Document()
+                source = Document.Meta.Source(
+                    table=table.table_name,
+                    row_id=row_id,
+                    scn=scn,
+                )
+                doc.meta_ = Document.Meta(op=op_mapping[op], source=source)
+                for (k, v) in row.items():
+                    if k in (scn_column_name, op_column_name, rowid_column_name):
+                        continue
+                    setattr(doc, k, v)
+
+                yield doc
+
+                # The query to fetch incremental changes uses "VERSIONS BETWEEN X and Y" where
+                # BETWEEN is an inclusive operator. So once we have processed all items of one SCN,
+                # in order to move on to the next SCN we need to increment the SCN by one. Since the query
+                # is always inclusive, this does not risk losing any subsequent events.
+                if first_transaction is None:
+                    first_transaction = scn
+                elif current_transaction is None and scn > first_transaction:
+                    current_transaction = scn
+                    yield scn + 1
+                elif current_transaction is not None and scn > current_transaction:
+                    yield current_transaction + 1
+                    current_transaction = scn
+
+    if first_transaction is not None and current_transaction is None:
+        yield first_transaction + 1
+    elif current_transaction is not None and current_transaction > scn:
+        yield current_transaction + 1
+
+    if full_state.backfill is not None:
```

Comment:
Is it actually necessary to check the backfill state here? As the early-return logic is written, if there's no backfill going on then any `sync_counter` value is fine. Alternatively the early-return logic doesn't need to check if the backfill state is `None` if the counter is going to stop incrementing while it's on an acceptable value for change streaming.

Reply:
@willdonnelly it is possible that we start the connector in a state in which it does not require a backfill at all (think: an empty table with no maximum rowid), in that case the sync counter will still start at 0...  we can probably in that scenario just set the sync counter to 1 and remove this condition, but I find more comfort in knowing that this function will definitely run if there is no backfill for any reason